tests/helpers/test_mkfds.c

```diff
@@ -3438,6 +3444,91 @@ static void *make_sockdiag(const struct factory *factory, struct fdesc fdescs[],
 	return NULL;
 }
 
+static void *make_foreign_sockets(const struct factory *factory _U_, struct fdesc fdescs[],
+				  int argc _U_, char ** argv _U_)
+{
+	int foreign_sd[2];	/* These two */
+	int original_ns;
+	int foreign_ns;
+	struct stat foreign_ns_sb;
+
+	original_ns = open("/proc/self/ns/net", O_RDONLY);
+	if (original_ns < 0)
+		err(EXIT_FAILURE, "failed in open(/proc/self/ns/net) before unshare");
+
+	if (unshare(CLONE_NEWNET) < 0) {
+		int e = errno;
+		close(original_ns);
+		errno = e;
+		err((errno == EPERM? EXIT_EPERM: EXIT_FAILURE),
+		    "failed in unshare()");
+	}
+
+	foreign_ns = open("/proc/self/ns/net", O_RDONLY);
+	if (foreign_ns < 0)
+		err(EXIT_FAILURE, "failed in open(/proc/self/ns/net) after unshare");
+	if (fstat(foreign_ns, &foreign_ns_sb) < 0) {
+			int e = errno;
+			close(foreign_ns);
+			close(original_ns);
+			errno = e;
+			err(EXIT_FAILURE, "failed in fstat(NETNS)");
+	}
```

Comment:
`goto` is not evil :-)
```
if (fstat(foreign_ns, &foreign_ns_sb) < 0) {
   warn("failed in fstat(NETNS)");
   goto fail;
}
...
fail:
   close(foreign_ns);
   close(original_ns);
   exit(EXIT_FAILURE);
```

Reply:
BTW, do we need close-before exit? Maybe just `err(EXIT_FAILURE, "failed in fstat(NETNS)");` is fine.

Reply:
See #3223.